src/Common/HostResolvePool.cpp

```diff
@@ -0,0 +1,293 @@
+{
+    update();
+}
+
+HostResolver::~HostResolver()
+{
+    std::lock_guard lock(mutex);
+    CurrentMetrics::sub(metrics.active_count, records.size());
+    records.clear();
+}
+
+void HostResolver::Entry::setFail()
+{
+    fail = true;
+
+    if (auto lock = pool.lock())
+        lock->setFail(address);
+}
+
+HostResolver::Entry::~Entry()
+{
+    if (!fail)
+    {
+        if (auto lock = pool.lock())
+            lock->setSuccess(address);
+    }
+}
+
+void HostResolver::update()
+{
+    MemoryTrackerSwitcher switcher{&total_memory_tracker};
+
+    auto next_gen = resolve_function(host);
+    if (next_gen.empty())
+        throw NetException(ErrorCodes::DNS_ERROR, "no endpoints resolved for host {}", host);
+
+    std::sort(next_gen.begin(), next_gen.end());
+
+    Poco::Timestamp now;
+
+    std::lock_guard lock(mutex);
+    updateImpl(now, next_gen);
+}
+
+void HostResolver::reset()
+{
+    std::lock_guard lock(mutex);
+
+    CurrentMetrics::sub(metrics.active_count, records.size());
+    records.clear();
+}
+
+void HostResolver::updateWeights()
+{
+    updateWeightsImpl();
+
+    if (getTotalWeight() == 0 && !records.empty())
+    {
+        for (auto & rec : records)
+            rec.failed = false;
+
+        updateWeightsImpl();
+    }
+
+    chassert((getTotalWeight() > 0 && !records.empty()) || records.empty());
+    random_weight_picker = std::uniform_int_distribution<size_t>(0, getTotalWeight() - 1);
+}
+
+HostResolver::Entry HostResolver::resolve()
+{
+    if (isUpdateNeeded())
+        update();
+
+    std::lock_guard lock(mutex);
+    return Entry(*this, selectBest());
+}
+
+void HostResolver::setSuccess(const Poco::Net::IPAddress & address)
+{
+    std::lock_guard lock(mutex);
+
+    auto it = find(address);
+    if (it == records.end())
+        return;
+
+    auto old_weight = it->getWeight();
+    ++it->usage;
+    auto new_weight = it->getWeight();
+
+    if (old_weight != new_weight)
+        updateWeights();
+}
+
+void HostResolver::setFail(const Poco::Net::IPAddress & address)
+{
+    Poco::Timestamp now;
+
+    {
+        std::lock_guard lock(mutex);
+
+        auto it = find(address);
+        if (it == records.end())
+            return;
+
+        it->failed = true;
+        it->fail_time = now;
+    }
+
+    ProfileEvents::increment(metrics.failed);
+    update();
+}
+
+Poco::Net::IPAddress HostResolver::selectBest()
+{
+    chassert(!records.empty());
+    size_t weight = random_weight_picker(thread_local_rng);
+    auto it = std::partition_point(records.begin(), records.end(), [&](const Record & rec) { return rec.weight_prefix_sum <= weight; });
+    chassert(it != records.end());
+    return it->address;
+}
+
+HostResolver::Records::iterator HostResolver::find(const Poco::Net::IPAddress & addr) TSA_REQUIRES(mutex)
+{
+    return std::lower_bound(
+        records.begin(), records.end(), addr, [](const Record & rec, const Poco::Net::IPAddress & value) { return rec.address < value; });
```

Comment:
Need to also check that it's actually equal.

Reply:
fixed